cmd/kubeadm/app/util/apiclient/idempotency.go

```diff
@@ -51,6 +54,46 @@ func CreateOrUpdateConfigMap(client clientset.Interface, cm *v1.ConfigMap) error
 	return nil
 }
 
+// CreateOrMutateConfigMap tries to create the ConfigMap provided as cm. If the resource exists already, the latest version will be fetched from
+// the cluster and cmmutator callback will be called on it, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the configmap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func CreateOrMutateConfigMap(client clientset.Interface, cm *v1.ConfigMap, cmmutator ConfigMapMutator) error {
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrap(err, "unable to create configmap")
+		}
+		return MutateConfigMap(client, metav1.ObjectMeta{Namespace: cm.ObjectMeta.Namespace, Name: cm.ObjectMeta.Name}, cmmutator)
+	}
+	return nil
+}
+
+// MutateConfigMap takes a ConfigMap Object Meta (namespace and name), retrieves the resource from the server and tries to mutate it
+// by calling to the cmmutator callback, then an Update of the mutated ConfigMap will be performed. This function is resilient
+// to conflicts, and a retry will be issued if the configmap was modified on the server between the refresh and the update (while the mutation was
+// taking place)
+func MutateConfigMap(client clientset.Interface, cmmeta metav1.ObjectMeta, cmmutator ConfigMapMutator) error {
+	for {
+		cm, err := client.CoreV1().ConfigMaps(cmmeta.Namespace).Get(cmmeta.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if err = cmmutator(cm); err != nil {
+			return errors.Wrap(err, "unable to mutate configmap")
+		}
+		_, err = client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(cm)
+		switch {
+		case err == nil:
+			return nil
+		case apierrors.IsConflict(err):
+			klog.V(2).Infof("conflict when updating %q configmap, retrying", cm.ObjectMeta.Name)
```

Comment:
1) what would be the retry rate i.e. how many calls per second?
does it make sense to add a small "wait" in the for loop?

2) do we want to cap the number of retries to a relatively high number?

Reply:
`configmap -> ConfigMap`

Reply:
Alternatively RetryOnConflict from client-go/util/retry could be used here: https://github.com/kubernetes/client-go/blob/master/util/retry/util.go#L61

Reply:
Ideally we want to use `PollImmediateInfinite` here. Due to the nature of how the API server works and how we use the config map, this will end at some point (when all other clients have the CM updated). However we want an interval value in order to avoid bringing down an API server upon large number of simultaneous updates.

Reply:
@neolit123 we are not consistent on `configmap` vs `ConfigMap` on `idempotency.go`. Will update to `ConfigMap` what this patch is changing, but we probably want to follow up to make everything consistent.